### PR TITLE
Add iotedgehubdev integration

### LINKED ISSUE #202: Integrate with iotedgehubdev
Currently there are 4 runtime control commands which need be ported for iotedgehubdev (for local development):
* `start`: wrapped `iotedgehubdev start -d <path-to-deployment-manifest>` and `iotedgehubdev start -i <input name>`, like `iotedgedev start` (pass deployment manifest path in .env as the `<path-to-deployment-manifest>` parameter or let users specify via CLI) and `iotedgedev start -i <input name>`
* `stop`: wrapped `iotedgehubdev stop`
* `setup`: `iotedgedev setup` wraps `iotedgehubdev setup -c <connection-string>` (pass device connection in .env as the parameter or let users specify via CLI)
* `status`: N/A, no `status` counterpart in iotedgehubdev yet.

### PR DESCRIPTION
**Pull request by LazarusX**
This PR closes #202

### CODE REVIEW

**Comment by jongio** [comment]
Please also update the container to includes iotedgehubdev instead of the old ctl.  Work with @yorek on coordinating that as he has a PR out to improve the way the containers are built.

@yorek - I'm thinking that instead of running the full agent in the container, we just run iotedgehubdev (simulator) - because the container is only dev scenarios anyway.

So, @LazarusX should be able to do a PR into your PR that just swaps the old ctl for iotedgehubdev.

**Comment by yorek** [comment]
Sounds good to me, I may need some help with iotedgehubdev, never used before.

**Comment by jongio** [comment]
@LazarusX - I get this when I run `iotedgedev simulator start`

```
ERROR: Error: [Errno 2] No such file or directory: 'C:\\ProgramData\\iotedgehubdev\\config\\edgehub.json'
ERROR: Error while executing command: iotedgehubdev start -d config\deployment.json. Command '['iotedgehubdev', 'start', '-d', 'config\\deployment.json']' returned non-zero exit status 1.
```
Looks like I need to run setup first, which I did, then got this:

```
ERROR: no such image: ${MODULES.filtermodule.amd64}: invalid reference format: repository name must be lowercase
ERROR: Error: Error while executing command: docker-compose -f C:\ProgramData\iotedgehubdev\data\data\docker-compose.yml up -d. Command '['docker-compose', '-f', 'C:\\ProgramData\\iotedgehubdev\\data\\data\\docker-compose.yml', 'up', '-d']' returned non-zero exit status 1.
ERROR: Error while executing command: iotedgehubdev start -d config\deployment.json. Command '['iotedgehubdev', 'start', '-d', 'config\\deployment.json']' returned non-zero exit status 1.

```

**Comment by jongio** [comment]
@LazarusX - Please also write some tests for this.  Thanks!

**Comment by jongio** [comment]
@LazarusX - Can runtime.py now be removed?

**Comment by LazarusX** [comment]
@yorek You can see a quickstart instruction for iotedgehubdev at https://pypi.org/project/iotedgehubdev/ (more detailed documentations are coming soon).

**Comment by LazarusX** [comment]
@jongio 

> @LazarusX - I get this when I run `iotedgedev simulator start`
> 
> ```
> ERROR: Error: [Errno 2] No such file or directory: 'C:\\ProgramData\\iotedgehubdev\\config\\edgehub.json'
> ERROR: Error while executing command: iotedgehubdev start -d config\deployment.json. Command '['iotedgehubdev', 'start', '-d', 'config\\deployment.json']' returned non-zero exit status 1.
> ```
Yes. You need to run setup before start to generate credentials. I am updating help messages to make it clear. 
> Looks like I need to run setup first, which I did, then got this:
> 
> ```
> ERROR: no such image: ${MODULES.filtermodule.amd64}: invalid reference format: repository name must be lowercase
> ERROR: Error: Error while executing command: docker-compose -f C:\ProgramData\iotedgehubdev\data\data\docker-compose.yml up -d. Command '['docker-compose', '-f', 'C:\\ProgramData\\iotedgehubdev\\data\\data\\docker-compose.yml', 'up', '-d']' returned non-zero exit status 1.
> ERROR: Error while executing command: iotedgehubdev start -d config\deployment.json. Command '['iotedgehubdev', 'start', '-d', 'config\\deployment.json']' returned non-zero exit status 1.
> 
> ```
It seems that you didn't build the solution before running based on the image name `${MODULES.filtermodule.amd64}`. You can run `iotedgedev build` or append `-b` parameter to `iotedgedev simulator start`.

**Comment by jongio** [comment]
@LazarusX - In start, can you automatically detect if setup hasn't been called and call it for the user?

In start can you detect that it hasn't been build and let them know.  The error wasn't enough to get me there.  Thanks, Jon

**Comment by LazarusX** [comment]
@jongio 
> @LazarusX - Can runtime.py now be removed?

I am removing it, as well as updating https://github.com/Azure/iotedgedev/issues/201 to document the removal of azure-iot-edge-ctl commands and addition of iotedgehubdev integration

**Comment by LazarusX** [comment]
@jongio 

> @LazarusX - In start, can you automatically detect if setup hasn't been called and call it for the user?

That's a bug with iotedgehubdev. I have filed issue https://github.com/Azure/iotedgehubdev/issues/69.

> In start can you detect that it hasn't been build and let them know. The error wasn't enough to get me there. Thanks, Jon

I have added the check.

**Comment by LazarusX** [comment]
@jongio I have add commits to resolve your comments. Please help review again.

File: tests/test_simulator.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    out, err = capfd.readouterr()
+    print(out)
+    print(err)
+    print(result.output)
+
+    if PY35:
+        assert 'Starting event monitor' in out
+    else:
+        assert 'Monitoring events from device' in out
+    assert 'timeCreated' in out
```
Comment (jongio) [inline tests/test_simulator.py]: Bad merge.  The last line should be removed.

**Comment by jongio** [comment]
Get the following with tox

```
=============================================
======== STOPPING IOT EDGE SIMULATOR ========
=============================================




.ERROR: Error: Error during pull for image hello-world:latest : 500 Server Error: Internal Server Error ("Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io on 192.168.65.1:53: read udp 192.168.65.3:50335->192.168.65.1:53: i/o timeout")
```

```
>       assert 'IoT Edge Simulator has been started in single module mode.' in result.output
E       assert 'IoT Edge Simulator has been started in single module mode.' in '\n============================================================\n======== STARTING IOT EDGE SIMULATOR IN SINGLE MODE ===...lookup mcr.microsoft.com on 192.168.65.1:53: read udp 192.168.65.3:35990->192.168.65.1:53: i/o timeout")\nERROR: \nNone\n'
E        +  where '\n============================================================\n======== STARTING IOT EDGE SIMULATOR IN SINGLE MODE ===...lookup mcr.microsoft.com on 192.168.65.1:53: read udp 192.168.65.3:35990->192.168.65.1:53: i/o timeout")\nERROR: \nNone\n' = <Result SystemExit()>.output

..\test_simulator.py:59: AssertionError
_____________________________ test_start_solution _____________________________

capfd = <_pytest.capture.CaptureFixture object at 0x000000000589C940>
```

```
>       assert 'IoT Edge Simulator has been started in solution mode.' in out
E       assert 'IoT Edge Simulator has been started in solution mode.' in 'ERROR: Error: Error during pull for image hello-world:latest : 500 Server Error: Internal Server Error ("Get https://...son. Command \'[\'iotedgehubdev\', \'start\', \'-d\', \'config\\deployment.json\']\' returned non-zero exit status 1\n\n'

..\test_simulator.py:97: AssertionError
________________________________ test_monitor _________________________________
```

**Comment by LazarusX** [comment]
@jongio That's weird. What's the result of running `docker run hello-world` on your machine? The hello-world image in iotedgehubdev is used as a helper to copy credentials to the Docker volume.

File: tests/test_simulator.py (hunk 1 above)
Comment (LazarusX) [reply tests/test_simulator.py]: The `tests/test_simulator.py` is added by me. The `timeCreated` line is used to check whether we can successfully send D2C messages using iotedgehubdev.

**Comment by jongio** [comment]
@LazarusX - I ran this:

iotedgedev solution create
iotedgedev build
iotedgedev push
iotedgedev deploy
iotedgedev start
and got this

```
Network azure-iot-edge-dev is external, skipping
ERROR: no such image: ${MODULES.filtermodule.amd64}: invalid reference format: repository name must be lowercase
ERROR: Error: Error while executing command: docker-compose -f C:\ProgramData\iotedgehubdev\data\data\docker-compose.yml up -d. Command '['docker-compose', '-f', 'C:\\ProgramData\\iotedgehubdev\\data\\data\\docker-compose.yml', 'up', '-d']' returned non-zero exit status 1.
ERROR: Error while executing command: iotedgehubdev start -d config\deployment.json. Command '['iotedgehubdev', 'start', '-d', 'config\\deployment.json']' returned non-zero exit status 1.

```

I checked config\deployment.json and see the ${} syntax in the file.

I rerun build

and then the config\deployment.json file has the right settings.


I traced it down.

For some reason, the 'deploy' step is not expanding the envvars to the config file correctly.

So run the steps above.  Run the deploy step and then look at config\deployment.json

File: tests/test_simulator.py (hunk 1 above)
Comment (jongio) [reply tests/test_simulator.py]: Thanks.  Please increase the timeout to more than 2 seconds as well.  Something like 15 seconds should be enough.

**Comment by LazarusX** [comment]
@jongio This should be fixed now. Please try again. I also added more commits.

**Comment by LazarusX** [comment]
Due to the issue I mentioned in https://github.com/Azure/iotedgedev/pull/236#discussion_r208494578, if tox stuck after starting the simulator in solution mode for Python 3.6, you might need to open another terminal and run `iotedgedev stop`.

**Comment by jongio** [comment]
For message:

 setup      Setup IoT Edge simulator. This must be done before starting

Does setup still need to be called before start?

**Comment by jongio** [comment]
@LazarusX - I'm seeing this message: 

image

**Comment by LazarusX** [comment]
Yes. This is an issue with iotedgehubdev https://github.com/Azure/iotedgehubdev/issues/69. Currently, iotedgehubdev does't provide a method to tell if the user has setup.

**Comment by LazarusX** [comment]
@jongio Conflict resolved.

**Comment by jongio** [comment]
Before I can merge this, we'll need to figure out how to kill monitor-process.  

Can you think of other options that would be a quick fix?  For example, can you call stop once you have verified that messages are received? Or set a timer and call stop when it expires?  Or only send a few messages?

I also have a branch where @digimaun and I are trying to figure out how to kill the process, but tox is not running there.  I'll work on that today. But a quick fix for you to unblock this merge would be ideal.

Jon

**Comment by LazarusX** [comment]
@jongio Do you mean temporarily getting this PR pass the tox e2e tests, or making the `timeout` option actually work as advertised on Python 3.6? For former, you can open another terminal window and run `iotedgedev stop` when you see tox get stuck (usually after creating the containers). For the latter, I might have to think how to implement it in my tomorrow morning :).

**Comment by jongio** [comment]
We need to be able to automate tox. So opening another window is not an option. Sure, get some sleep :)

**Pull request merged**